forte/orbital-helpers/aosubspace.h

```diff
@@ -127,58 +131,51 @@ class AOInfo {
  *    ["C(2p)"] - the 2p subset of all carbon atoms
  *    ["C(1s)","C(2s)"] - the 1s/2s subsets of all carbon atoms
  *    ["C1-3(2s)"] - the 2s subsets of carbon atoms #1, #2, #3
+ *    ["Ce(4fzx2-zy2)"] - the 4f zxx-zyy orbital of all Ce atoms
  */
 class AOSubspace {
   public:
     // ==> Constructors <==
 
-    // Simple constructor
-    AOSubspace(std::shared_ptr<psi::Molecule> molecule, std::shared_ptr<psi::BasisSet> basis);
-    // Constructor with list of subspaces
+    /// Constructor using a list of subspaces and atom normals
+    /// @param subspace_str: a list of subspace orbitals, e.g, {"C2", "N", "Fe(3d)", "Mo(3dx2-y2)"}
+    /// @param molecule: a Psi4 Molecule object
+    /// @param basis: a Psi4 basis object, a minimal basis where subspace orbitals are selected
+    /// @param atom_normals: (optional) a map from the atom to its normal
+    /// @param debug_mode: debug mode if True (more printing)
+    ///
+    /// For the argument of atom_normals:
+    ///   An atom is characterized by its atomic number and the relative index.
+    ///   For example, {6, 2} - the third carbon atom of the molecule (index is 0-based)
+    ///
+    ///   This argument is used to make the subspace p orbitals of an atom aligned to the normal:
+    ///   px, py, pz (3 orbitals) -> nx * px + ny * py + nz * pz (1 orbital),
+    ///   where the atom normal is a 3D unit vector (nx, ny, nz).
+    ///
+    ///   If the map is empty, the p orbitals are in the xyz frame defined the molecule.
+    ///   Equivalently, 3 normals are attached to each atom: (1,0,0), (0,1,0), and (0,0,1).
+    ///   As such, we still have a full set of p orbitals in the subspace.
     AOSubspace(std::vector<std::string> subspace_str, std::shared_ptr<psi::Molecule> molecule,
-               std::shared_ptr<psi::BasisSet> basis);
+               std::shared_ptr<psi::BasisSet> basis,
+               std::map<std::pair<int, int>, psi::Vector3> atom_normals = {},
```

Comment:
The choice of passing the `pair<int,int>` implies some sort of processing inside the class, where one then has to find which atom maps to which atom in the molecule. I understand that this might take extra time, but a consistent solution is to pass a map<std::vector<std::string>,psi::Vector3> ( for example {["C","H","O"] : (1.0, 0.0,0.0)} ).
Is that hard to do?

Reply:
I use `pair<int, int>` to **prevent** any processing of `atom_normals` inside the `AOSubspace` class, because `atom_to_aos_` is a vector of vector, i.e, atomic Z to a list of AO number. The processing is done outside the class.

The loop structure of `AOSubspace` is looping over the subspace orbitals, e.g., `{"C(2p)", "O(2p)"}`. For each entry (e.g., `"C(2p)"`), loop over the atom indices in the molecule. For each atom, loop over the AOs and push the matching AO to subspace.
I only need to add a few `if`s to make the atom normals to fit in this loop structure.

Although I am not a big fan of the current data structure, changing this would mean an overhaul to the code.